Plugin system

asutoshpalai opened the pull request:
This is the working PR for plugin system for bundler.

A rough workflow shall be
- [x] Add functionality to install a plugin as gem from RubyGem source and also from git source
- [x] Add support for command line plugin
- [ ] Add support for source plugin
- [ ] Add support for life cycle hooks
- [ ]  Finalize all the functionalities

segiddins commented on lib/bundler/cli/plugin.rb:
```diff
@@ -0,0 +1,23 @@
+    method_option "source", :type => :string, :default => nil, :banner =>
+      "URL of the RubyGems source to fetch the plugin from"
+    method_option "version", :type => :string, :default => nil, :banner =>
+      "The version of the plugin to fetch from"
```
~~from~~

segiddins commented on lib/bundler/dsl.rb:
```diff
@@ -19,7 +19,7 @@ def self.evaluate(gemfile, lockfile, unlock)
     def initialize
       @source               = nil
-      @sources              = SourceList.new
+      @sources            ||= SourceList.new
```
why is this change necessary?

segiddins commented on lib/bundler/plugin.rb:
```diff
@@ -0,0 +1,89 @@
+    autoload :Installer,  "bundler/plugin/installer"
+    autoload :SourceList, "bundler/plugin/source_list"
+
+    class << self
```
since this class can't be instantiated, maybe it should be a module?

segiddins commented on lib/bundler/plugin/dsl.rb:
```diff
@@ -0,0 +1,27 @@
+
+    # So that we don't have to override all there methods to dummy ones
+    # explicitly.
+    # They will be handled by missing_methods
```
`method_missing`

segiddins commented on lib/bundler/plugin/dsl.rb:
```diff
@@ -0,0 +1,27 @@
+    def initialize
+      @sources = Plugin::SourceList.new
+
+      super
```
`super` should be called first

segiddins commented on lib/bundler/source.rb:
```diff
@@ -7,14 +7,22 @@ class Source
     attr_accessor :dependency_names
 
+    def initialize
+      @options = {}
```
this is a bit of a code smell to me

segiddins commented:
If at all possible, it'd be great to get rid of all of the `if @options[:plugin]` -- in my eyes, it looks like that's introducing a look of branching where it may not be entirely necessary

asutoshpalai commented:
`@options[:plugin]` stemmed out of my attempt to use the existing source classes for fetch and install the plugins.

While writing that, I tired real hard to avoid `@options[:plugin]`. But the alternate solutions leads to a lot of code duplication or a major refactor of the the source classes.

I am still in lookout for solution and would definitely appreciate a few pointers.

I expect that development of source plugins might show me some way out of it, but research on those lines up till doesn't seem to thought any light in this direction.

asutoshpalai commented:
@segiddins : fixed the issues you pointed out (including the code smell :smile_cat: ). Removed `@options[:plugin]`.

The command function is also ready.

segiddins commented on lib/bundler/cli/install.rb:
```diff
@@ -77,25 +49,15 @@ def run
       # rubygems plugins sometimes hook into the gem install process
       Gem.load_env_plugins if Gem.respond_to?(:load_env_plugins)
 
+      Plugin.eval_gemfile(Bundler.default_gemfile) if ENV["BUNDLE_PLUGIN"]
```
it looks like this is being called for side effects, which is confusing

asutoshpalai replied:
I was not able to find any better place for it.

Can you please suggest some

segiddins replied:
well `eval_gemfile` suggests it will return something, if it doesn't then that method should be renamed

segiddins commented on lib/bundler/cli/install.rb:
```diff
@@ -190,5 +154,64 @@ def check_for_options_conflicts
+      Bundler.settings[:path]     = nil if options[:system]
+      Bundler.settings[:path]     = "vendor/bundle" if options[:deployment]
+      Bundler.settings[:path]     = options["path"] if options["path"]
+      Bundler.settings[:path] ||= "bundle" if options["standalone"]
```
the alignment in this method is now off
also, it'd be nice to have a blank line between the different settings?

segiddins commented on lib/bundler/cli/plugin.rb:
```diff
@@ -0,0 +1,23 @@
+      "The git branch to checkout"
+    method_option "ref", :type => :string, :default => nil, :banner =>
+      "The git revision to check out"
+    def install(plugin)
```
any reason this couldn't install multiple plugins at once?

asutoshpalai replied:
Only that, I was not able to come up with some flow to accept the version/ref for multiple plugins.

segiddins replied:
use the same for all of them?

asutoshpalai replied:
or use the default and raise error when multiple plugins are specified and a version is given?

segiddins replied:
nah since you might want multiple gems at the same version, like if they're all in a suite of gems

segiddins commented on lib/bundler/plugin.rb:
```diff
@@ -0,0 +1,131 @@
+    autoload :Installer,  "bundler/plugin/installer"
+    autoload :SourceList, "bundler/plugin/source_list"
+
+    PLUGIN_FILE_NAME = "plugin.rb".freeze
```
`plugins.rb`?

segiddins commented on lib/bundler/plugin.rb:
```diff
@@ -0,0 +1,131 @@
+
+    @commands = {}
+
+    class << self
```
since this is never instantiated, maybe it should be a module with `module_function` instead?

asutoshpalai replied:
`module_function` doesn't provide the ability to define private methods on the module. That leaves us with two options: 1. Give up on singleton pattern for Plugin, 2. Make the private methods public.

I am aware that [bbatsov/ruby-style-guide](https://github.com/bbatsov/ruby-style-guide#module-function) recomends it, but I don't find it reflected in [GitHub's](https://github.com/airbnb/ruby) or [Airbnb's](https://github.com/styleguide/ruby)

segiddins commented on lib/bundler/plugin/api.rb:
```diff
@@ -0,0 +1,54 @@
+  # To use it without inheriting, you will have to create an object of this
+  # to use the functions (except for declaration functions like command, source,
+  # and hooks).
+  class Plugin::Api
```
`API`

segiddins commented on lib/bundler/plugin/dsl.rb:
```diff
@@ -0,0 +1,26 @@
+
+module Bundler
+  # Dsl to parse the Gemfile looking for plugins to install
+  class Plugin::Dsl < Bundler::Dsl
```
`DSL`

segiddins commented on lib/bundler/plugin/dsl.rb:
```diff
@@ -0,0 +1,26 @@
+    end
+
+    def method_missing(name, *args)
+      # Dummy evaluation
```
can we please only ignore methods defined on `Bundler::Dsl` ? otherwise this could lead to bugs in this class

segiddins commented on lib/bundler/plugin/index.rb:
```diff
@@ -0,0 +1,73 @@
+  # Manages which plugins are installed and their sources. This also is supposed to map
+  # which plugin does what (currently the features are not implemented so this class is
+  # now a stub class).
+  class Plugin::Index
```
please nest the nested class

segiddins commented on lib/bundler/plugin/index.rb:
```diff
@@ -0,0 +1,73 @@
+      @plugin_paths[name] = path
+
+      common = commands & @commands.keys
+      raise "Command(s) #{common.join(", ")} are already registered" if common.any?
```
this should raise a specific error

segiddins commented on lib/bundler/plugin/index.rb:
```diff
@@ -0,0 +1,73 @@
+      SharedHelpers.filesystem_access(index_file, :read) do |index_f|
+        valid_file = index_f && index_f.exist? && !index_f.size.zero?
+        break unless valid_file
+        index = YAML.load_file(index_f)
```
if this can _ever_ be used in the same process as `bundle exec`, we can't use YAML

segiddins commented on lib/bundler/plugin/installer.rb:
```diff
@@ -0,0 +1,102 @@
+      paths = {}
+
+      specs.each do |spec|
+        next if spec.name == "bundler"
```
is this really necessary?

asutoshpalai replied:
I can't remember why I wrote this, maybe some reminiscent of some other installation flow I was trying out.

Anyways it is not necessary now, so removing it

homu commented:
:umbrella: The latest upstream changes (presumably #4617) made this pull request unmergeable. Please resolve the merge conflicts.

segiddins commented on lib/bundler/plugin/index.rb:
```diff
@@ -23,7 +30,7 @@ def register_plugin(name, path, commands)
 
         common = commands & @commands.keys
-        raise "Command(s) #{common.join(", ")} are already registered" if common.any?
+        raise CommandConflict.new(name, common) if common.any?
```
`unless common.empty?`

segiddins commented on lib/bundler/cli.rb:
```diff
@@ -82,6 +82,10 @@ def help(cli = nil)
     end
 
     def self.handle_no_command_error(command, has_namespace = $thor_runner)
+      if ENV["BUNDLE_PLUGIN"] && Bundler::Plugin.command?(command)
```
could this please using `Bundler.settings`?

segiddins commented on lib/bundler/cli.rb:
```diff
@@ -437,6 +441,12 @@ def env
       Env.new.write($stdout)
     end
 
+    if ENV["BUNDLE_PLUGIN"]
```
settings

asutoshpalai replied:
I did that in  4a1baea, but some specs were failing because when called here (during the loading of `Cli` class)  `Bundler.settings` gets initialized before `CLI` class gets initialized.

So some options like `--gemfile` don't get properly set. Initializing `Settings` calls `Bundler.root` which caches the root path determined without respecting `--gemfile` option (which is set in [cli#initialize](https://github.com/bundler/bundler/blob/master/lib/bundler/cli.rb#L21)).

So reverted it cf284c0. I tried digging but failed to fix the problem.

asutoshpalai replied:
We can allow `plugin` sub command and show message if called but not enabled.

segiddins commented on lib/bundler/plugin/api.rb:
```diff
@@ -0,0 +1,56 @@
+      # @param [String] name unique for the plugin or the purpose
+      # @return [Pathname] object for the new directory created
+      def tmp(name)
+        Bundler.tmp(File.join(["plugin", name]))
```
I think this can just be `Bundler.tmp("plugins", name)`

asutoshpalai replied:
`Bundler.tmp` doesn't accept splatted arguments. But :+1: on making name splatted, we can use `join`

segiddins commented on lib/bundler/plugin/index.rb:
```diff
@@ -0,0 +1,116 @@
+        end
+      end
+
+      def dump_yaml(hash)
```
the YAML stuff should be refactored so it can be shared with settings, and unit tested

asutoshpalai replied:
I had already refactored it, just hadn't pushed it yet. I was trying to make the specs pass with `Settings` using it.
Made the specs pass :grinning:

segiddins commented on lib/bundler/plugin/installer.rb:
```diff
@@ -0,0 +1,99 @@
+          source = options[:source]
+          install_rubygems(names, version, source)
+        else
+          raise(ArgumentError, "You need to provide the source")
```
shouldn't it default to `Gem.sources` if no source is explicitly provided?

segiddins commented:
This is looking really good! Just for fun, could you try adding a spec for using plugins in an inline gemfile?

asutoshpalai commented:
Added thee default gem source, spec for inline gemfile and shifted settings to the new serializer. 

The only thing that remains is use of `ENV` as commented above. I believe that we should allow the sub command and show message when used.

asutoshpalai commented:
Using `Bundler.settings` now :dart:

segiddins commented:
Cool! Can you just make sure `plugins` is now a bool key in Settings?

segiddins commented on lib/bundler/plugin.rb:
```diff
@@ -0,0 +1,156 @@
+      save_plugins paths
+    rescue PluginError => e
+      paths.values.map {|path| Bundler.rm_rf(path) } if paths
+      Bundler.ui.error "Failed to install plugin #{name}: #{e.message}\n  #{e.backtrace.join("\n  ")}"
```
here only the first  line of the backtrace will be indented

segiddins commented on lib/bundler/plugin.rb:
```diff
@@ -0,0 +1,156 @@
+      else
+        definition = DSL.evaluate(gemfile, nil, {})
+      end
+      return unless definition.dependencies.any?
```
`return if definition.dependencies.empty?`

segiddins commented on lib/bundler/plugin.rb:
```diff
@@ -0,0 +1,156 @@
+    # To be called from Cli class to pass the command and argument to
+    # approriate plugin class
+    def exec_command(command, args)
+      raise UndefinedCommandError, "Command #{command} not found" unless command? command
```
the interpolated command in backticks

segiddins commented on lib/bundler/plugin.rb:
```diff
@@ -0,0 +1,156 @@
+    # @raise [Error] if plugins.rb file is not found
+    def validate_plugin!(plugin_path)
+      plugin_file = plugin_path.join(PLUGIN_FILE_NAME)
+      raise MalformattedPlugin, "#{PLUGIN_FILE_NAME} was not found in the plugin!" unless plugin_file.file?
```
no `!`

segiddins commented on lib/bundler/plugin/api.rb:
```diff
@@ -0,0 +1,56 @@
+        Bundler.tmp(["plugin", *names].join("-"))
+      end
+
+      def method_missing(name, *args, &blk)
```
should also override `respond_to_missing?`

segiddins commented:
This gets a 👍 from me to ship in 1.13 (but disabled by default). There are some scenarios I'd like to see tested better (plugins with dependencies for example), but that can be done in a later PR. Well done, @asutoshpalai! @fotanus, I'm going to merge this by EOD unless you have any objections?

fotanus commented:
I haven't check the code closely enough, but I'm with you just because you are the most able person to do such decisions! 

The design is quite good by what I saw in the tests, and deviate little from the original proposal, so I think we are on point.

segiddins commented:
@homu r+

homu commented:
:pushpin: Commit 0aaa3f3 has been approved by `segiddins`

<!-- @homu r=segiddins 0aaa3f3a9b236acfc5198196c845cc9d6320ce30 -->

homu commented:
:hourglass: Testing commit 0aaa3f3 with merge fb72f36...

homu commented:
:sunny: Test successful - [status](https://travis-ci.org/bundler/bundler/builds/136821342)
